modules/Bio/EnsEMBL/Hive/HivePipeline.pm

```diff
@@ -544,151 +557,197 @@ sub print_diagram {
                 if( $attrib_name eq 'wait_for' ) {
-
+                    $tweakStructure->{Return}->{Field} = 'wait_for';
                     my $cr_collection   = $self->collection_of( 'AnalysisCtrlRule' );
                     my $acr_collection  = $analysis->control_rules_collection;
-
+                    $tweakStructure->{Return}->{OldValue} = join(', ', map { $_->condition_analysis_url } @$acr_collection );
                     if($operator eq '?') {
-                        print "Tweak.Show    \tanalysis[$analysis_name].wait_for ::\t[".join(', ', map { $_->condition_analysis_url } @$acr_collection )."]\n";
+                        $tweakStructure->{Action} = "SHOW";
+                        $tweakStructure->{Return}->{NewValue} = $tweakStructure->{Return}->{OldValue};
+                        push @response, "Tweak.Show  \tanalysis[$analysis_name].wait_for ::\t[".join(', ', map { $_->condition_analysis_url } @$acr_collection )."]\n";
                     }
 
                     if($operator eq '#' or $operator eq '=') {     # delete the existing rules
+                        $tweakStructure->{Action} = "DELETE";
+                        $tweakStructure->{Return}->{NewValue} = undef;
                         foreach my $c_rule ( @$acr_collection ) {
                             $cr_collection->forget_and_mark_for_deletion( $c_rule );
                             $need_write = 1;
 
-                            print "Tweak.Deleting\t".$c_rule->toString." --> (missing value)\n";
+                            push @response, "Tweak.Deleting\t".$c_rule->toString." --> (missing value)\n";
                         }
                     }
 
                     if($operator eq '=' or $operator eq '+=') {     # create new rules
+                        $tweakStructure->{Action} = "SET";
+                        $tweakStructure->{Return}->{NewValue} = $new_value;
                         Bio::EnsEMBL::Hive::Utils::PCL::parse_wait_for($self, $analysis, $new_value);
                         $need_write = 1;
                     }
 
                 } elsif( $attrib_name eq 'flow_into' ) {
-
+                    $tweakStructure->{Return}->{Field} = 'flow_into';
+                    $tweakStructure->{Return}->{OldValue} = $analysis->logic_name;
                     if($operator eq '?') {
                         # FIXME: should not recurse
-                        $analysis->print_diagram_node($self, '', {});
+                        $tweakStructure->{Action} = 'SHOW';
+                        $tweakStructure->{Return}->{NewValue} = $tweakStructure->{Return}->{OldValue};
+                        #$analysis->print_diagram_node($self, '', {});
                     }
 
                     if($operator eq '#' or $operator eq '=') {     # delete the existing rules
+                        $tweakStructure->{Action} = 'DELETE';
+                        $tweakStructure->{Return}->{NewValue} = undef;
```

Comment:
Since there wouldn't be _OldValue_ at all maybe we can skip setting _NewValue_ here

Reply:
Should set it to guarantee fields